document/json/encoder.go

```diff
@@ -0,0 +1,318 @@
+package json
+
+import (
+	"fmt"
+	"math/big"
+	"reflect"
+
+	"github.com/aws/smithy-go/document"
+	"github.com/aws/smithy-go/document/internal/serde"
+	smithyjson "github.com/aws/smithy-go/encoding/json"
+)
+
+type EncoderOptions struct{}
+
+type Encoder struct {
+	options EncoderOptions
+}
+
+func (e *Encoder) Encode(v interface{}) ([]byte, error) {
+	if document.IsNoSerde(v) {
+		return nil, fmt.Errorf("unsupported type: %v", v)
+	}
+
+	encoder := smithyjson.NewEncoder()
+
+	if err := e.encode(jsonValueProvider(encoder.Value), reflect.ValueOf(v), serde.Tag{}); err != nil {
+		return nil, err
+	}
+
+	encodedBytes := encoder.Bytes()
+
+	if len(encodedBytes) == 0 {
+		return nil, nil
+	}
+
+	return encodedBytes, nil
+}
+
+type jsonValueProvider smithyjson.Value
+
+func (p jsonValueProvider) GetValue() smithyjson.Value {
+	return smithyjson.Value(p)
+}
+
+type valueProvider interface {
+	GetValue() smithyjson.Value
+}
+
+type jsonObjectKeyProvider struct {
+	Object *smithyjson.Object
+	Key    string
+}
+
+type jsonArrayProvider struct {
+	Array *smithyjson.Array
+}
+
+func (p jsonArrayProvider) GetValue() smithyjson.Value {
+	return p.Array.Value()
+}
+
+func (p jsonObjectKeyProvider) GetValue() smithyjson.Value {
+	return p.Object.Key(p.Key)
+}
+
+func (e *Encoder) encode(vp valueProvider, rv reflect.Value, tag serde.Tag) error {
+	// Zero values are serialized as null, or skipped if omitEmpty.
+	if serde.IsZeroValue(rv) {
+		if tag.OmitEmpty {
+			return nil
+		}
+		return e.encodeZeroValue(vp, rv)
+	}
+
+	// Handle both pointers and interface conversion into types
+	rv = serde.ValueElem(rv)
+
+	switch rv.Kind() {
+	case reflect.Invalid:
+		if tag.OmitEmpty {
+			return nil
+		}
+		vp.GetValue().Null()
+		return nil
+
+	case reflect.Struct:
+		return e.encodeStruct(vp, rv)
+
+	case reflect.Map:
+		return e.encodeMap(vp, rv)
+
+	case reflect.Slice, reflect.Array:
+		return e.encodeSlice(vp, rv)
+
+	case reflect.Chan, reflect.Func, reflect.UnsafePointer:
+		// skip unsupported types
+		return nil
+
+	default:
+		return e.encodeScalar(vp, rv)
+	}
+}
+
+func (e *Encoder) encodeZeroValue(vp valueProvider, rv reflect.Value) error {
+	switch rv.Kind() {
+	case reflect.Invalid:
+		vp.GetValue().Null()
```

Comment:
I think this duplicates the `Invalid` case in `encode`. the case in `encode` probably isn't needed?

Reply:
See my reply to your earlier comment with regards to how Invalid is valid in both these cases.